src/CommandDialog.cpp

```diff
@@ -0,0 +1,161 @@
+#include "ScopedAssign.h"
+#include "CommandDialog.h"
+#include <QStandardItemModel>
+#include <map>
+
+
+enum CommandColumns {
+    ICON = 0, NAME, DESCRIPTION, COMMAND, HIDDEN, INDEX, MAX
+};
+
+
+CommandDialog::CommandDialog(QMap<QString, CommandRef>& commands, QWidget* parent)
+    : QDialog(parent), commands(commands)
+{
+    setupUi(this);
+
+    twCommands->resizeColumnsToContents();
+    twCommands->setColumnHidden(HIDDEN, true);
+    twCommands->verticalHeader()->hide();
+
+    connect(btnAdd, &QPushButton::clicked, this, &CommandDialog::onAddButtonClicked);
+    connect(btnRemove, &QPushButton::clicked, this, &CommandDialog::onRemoveButtonClicked);
+    connect(twCommands, &QTableWidget::itemChanged, this, &CommandDialog::changeTableItem);
+
+    connect(okButton, &QPushButton::clicked, this, &CommandDialog::accept);
+    connect(cancelButton, &QPushButton::clicked, this, &CommandDialog::reject);
+
+    refresh();
+}
+
+void CommandDialog::refresh()
+{
+    QList<QPair<CommandRef, QString>> list;
+
+    // set a comparison list
+    for (auto key: commands.keys()) {
+        list.append(QPair<CommandRef, QString>(commands[key], key));
+    }
+
+    // sort by index
+    std::sort(std::begin(list), std::end(list));
+    auto sa = ScopedAssign(userMode, false);
+
+    // populate
+    for (auto& pair: list) {
+        createItem();
+        writeCommand(pair.first);
+    }
```

Comment:
This is fine. I just want to mention the following possibility. Pick whatever you prefer:
```
for (const auto& [key, _] : list) {     // or a more descriptive name than 'key'
    ...
    writeCommand(key);
}
```